ext/cache/src/main.rs

```diff
@@ -0,0 +1,232 @@
+            generate_otp(&ctx, &algorithm, slot.id)?;
+            return Ok(());
+          }
+        }
+        anyhow::bail!("No OTP slot with the given name!");
+      }
+      OtpCommand::List => {
+        let cache: OtpCache = get_cache(&cache_file)?;
+        println!("alg\tslot\tname");
+        for (algorithm, slots) in cache {
+          for slot in slots {
+            println!("{}\t{}\t{}", algorithm, slot.id, slot.name);
+          }
+        }
+      }
+      OtpCommand::Update => {
+        let data = get_otp_slots(&ctx)?;
+        save_cache(&data, &cache_file)?;
+      }
+    },
+    Command::Pws(cmd) => match cmd {
+      PwsCommand::GetLogin { name } => {
+        let cache: PwsCache = get_cache(&cache_file)?;
+        if let Some(slot) = cache.slots.iter().find(|s| &s.name == name) {
+          get_pws_data(&ctx, slot.id, "--login")?;
+        } else {
+          anyhow::bail!("No PWS slot with the given name!");
+        }
+      }
+      PwsCommand::GetPassword { name } => {
+        let cache: PwsCache = get_cache(&cache_file)?;
+        if let Some(slot) = cache.slots.iter().find(|s| &s.name == name) {
+          get_pws_data(&ctx, slot.id, "--password")?;
+        } else {
+          anyhow::bail!("No PWS slot with the given name!");
+        }
+      }
+      PwsCommand::List => {
+        let cache: PwsCache = get_cache(&cache_file)?;
+        println!("slot\tname");
+        for slot in cache.slots {
+          println!("{}\t{}", slot.id, slot.name);
+        }
+      }
+      PwsCommand::Update => {
+        let data = get_pws_slots(&ctx)?;
+        save_cache(&data, &cache_file)?;
+      }
+    },
+  }
+
+  Ok(())
+}
+
+fn get_cache<T: serde::de::DeserializeOwned>(file: &path::Path) -> anyhow::Result<T> {
+  if !file.is_file() {
+    anyhow::bail!("There is no cached slot data.  Run the update command to initialize the cache.");
+  }
+  load_cache(&file)
+}
+
+fn load_cache<T: serde::de::DeserializeOwned>(path: &path::Path) -> anyhow::Result<T> {
+  let s = fs::read_to_string(path).context("Failed to read cache file")?;
+  toml::from_str(&s).context("Failed to parse cache file")
+}
+
+fn save_cache<T: serde::Serialize>(cache: &T, path: &path::Path) -> anyhow::Result<()> {
+  use io::Write as _;
+
+  if let Some(parent) = path.parent() {
+    fs::create_dir_all(parent).context("Failed to create cache parent directory")?;
+  }
+  let mut f = fs::File::create(path).context("Failed to create cache file")?;
+  f.write_all(&toml::to_vec(cache).context("Failed to serialize cache")?)
+    .context("Failed to write cache file")?;
+  Ok(())
+}
+
+fn get_serial_number(ctx: &ext::Context) -> anyhow::Result<String> {
+  let status = ext::Nitrocli::from_context(ctx)
+    .arg("status")
+    .text()
+    .context("Failed to query device status")?;
+  let r = regex::Regex::new(r#"(?m)^\s*serial number:\s*(\S.*)$"#)
+    .context("Failed to compile serial regex")?;
```

Comment:
That’s the bigger issue I mentioned previously. :-)

I made a draft supporting multiple output formats, but it still requires some work.  I agree that this is not optimal, but it relies on the format tested in our unit tests.

I dislike calling `nitrokey-rs` directly because it requires us:
- to know all options that affect the connection
- to re-implement parsing these options (e. g. model)
- to re-implement the selection of the correct device

If we would decide to do something like this, then I’d prefer doing it in a crate that can be shared between nitrocli and its extensions, but I recall you didn’t want to do that.

Reply:
Here is the draft for output formats: https://github.com/robinkrahl/nitrocli/tree/output-formats